### PR TITLE
Gracefully handle kernel cannot create livy session in wrapper kernels.

### PR DESCRIPTION
**Pull request by aggFTW**
Kernel will display a fatal failure message and will refuse to run code afterwards.

### CODE REVIEW

File: remotespark/livyclientlib/livysession.py [hunk 1]
```diff
@@ -68,13 +68,21 @@ def create_sql_context(self):
...
 
-        self.wait_for_status(Constants.idle_session_status, self._create_sql_context_timeout_seconds)
-        self.execute(self._get_sql_context_creation_command())
-        self.logger.debug("Started '{}' sql session.".format(self.language))
-
-        self.wait_for_status(Constants.idle_session_status, self._create_sql_context_timeout_seconds)
-        self.execute(self._get_hive_context_creation_command())
-        self.logger.debug("Started '{}' hive session.".format(self.language))
+        try:
+            self.wait_for_status(Constants.idle_session_status, self._create_sql_context_timeout_seconds)
```
Comment (msftristew) [inline remotespark/livyclientlib/livysession.py]: Will this method `wait_for_status` ever be called with a session status that is not `idle`? If so then you can consider changing the method to `wait_for_idle` and just doing away with the first argument

File: remotespark/livyclientlib/livysession.py [hunk 2]
```diff
@@ -68,13 +68,21 @@ def create_sql_context(self):
...
+            self.execute(self._get_sql_context_creation_command())
+            self.logger.debug("Started '{}' sql session.".format(self.language))
+        except LivyClientTimeoutError:
+            raise LivyClientTimeoutError("Failed to create the SQL context in time. Timed out after {} seconds."
+                                         .format(self._create_sql_context_timeout_seconds))
+
+        try:
+            self.wait_for_status(Constants.idle_session_status, self._create_sql_context_timeout_seconds)
```
Comment (msftristew) [inline remotespark/livyclientlib/livysession.py]: Copy-paste here. This and the above try-except block should probably be refactored into a separate method

File: remotespark/sparkkernelbase.py [hunk 3]
```diff
@@ -81,19 +60,74 @@ def do_execute(self, code, silent, store_history=True, user_expressions=None, al
...
         return super(SparkKernelBase, self).do_execute(code, silent, store_history, user_expressions, allow_stdin)
 
-    def do_shutdown_ipykernel(self, restart):
+    def _do_shutdown_ipykernel(self, restart):
         return super(SparkKernelBase, self).do_shutdown(restart)
+
+    def _abort_with_fatal_error(self, message):
+        self._fatal_error = message
+
+        error = get_configuration(Constants.fatal_error_suggestion, self.fatal_error_suggestion).format(message)
```
Comment (msftristew) [inline remotespark/sparkkernelbase.py]: From a design perspective I'd say it's more sensible for default values to be handled by the get_configuration function (or, maybe better, a Configuration class).  Having default values for configurations be spread out through the entire codebase seems messy.

File: remotespark/sparkkernelbase.py [hunk 4]
```diff
@@ -28,46 +35,18 @@ def __init__(self, **kwargs):
...
-        self.already_ran_once = True
-
     def do_execute(self, code, silent, store_history=True, user_expressions=None, allow_stdin=False):
+        if self._fatal_error is not None:
+            self._abort_with_fatal_error(self._fatal_error)
+
         if not self.already_ran_once:
-            (username, password, url) = self.get_configuration()
-            self.initialize_magics(username, password, url)
+            (username, password, url) = self._get_configuration()
```
Comment (msftristew) [inline remotespark/sparkkernelbase.py]: Idea: the _get_configuration method is only called once, here, and its output is passed directly to the _initialize_magics method. So maybe _initialize_magics should take no arguments and call _get_configuration itself.

**Comment by msftristew** [comment]
Made some comments.

File: remotespark/livyclientlib/livysession.py (hunk 1 above)
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: I don't know. We can make that change later.

File: remotespark/livyclientlib/livysession.py (hunk 2 above)
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: Done.

File: remotespark/sparkkernelbase.py (hunk 3 above)
Comment (aggFTW) [inline remotespark/sparkkernelbase.py]: We can do this later in one pass through the code.

File: remotespark/sparkkernelbase.py (hunk 4 above)
Comment (aggFTW) [inline remotespark/sparkkernelbase.py]: This was done for unit testing purposes.

**Pull request merged**